File: supabase/functions/_shared/logger.ts
// supabase/functions/_shared/logger.ts

/**
 * Centralized logger for Supabase Edge Functions
 * Provides structured logging with emoji indicators and request correlation
 * 
 * ## Usage
 * 
 * ```typescript
 * import { createLogger } from './_shared/logger.ts';
 * 
 * serve(async (req: Request) => {
 *   const logger = createLogger('my-function');
 *   logger.info('Function invoked');
 * });
 * ```
 * 
 * ## Log Levels
 * 
 * - `debug`: Detailed debugging information
 * - `info`: Important events (success, completion)
 * - `warn`: Potential issues (RLS blocks, validation warnings)
 * - `error`: Actual errors (exceptions, failed operations)
 * 
 * ## Emoji Legend
 * 
 * | Emoji | Meaning |
 * |-------|---------|
 * | 🔐 | Authentication |
 * | 📡 | Database query |
 * | 🔍 | RLS check |
 * | ⚡ | Edge function |
 * | ✅ | Success |
 * | ❌ | Error |
 * | ⚠️ | Warning |
 * | 🚫 | Blocked/Denied |
 * | 💥 | Critical failure |
 * | 👤 | User context |
 */

export interface LogMeta {
  [key: string]: any;
}

export interface EdgeLogger {
  debug: (message: string, meta?: LogMeta) => void;
  info: (message: string, meta?: LogMeta) => void;
  warn: (message: string, meta?: LogMeta) => void;
  error: (message: string, meta?: LogMeta) => void;
  setRequestId: (requestId: string) => void;
  setUserId: (userId: string) => void;
}

/**
 * Create a logger instance for an Edge Function
 * @param functionName Name of the edge function (e.g., 'toggle-recipe-like')
 */
export function createLogger(functionName: string): EdgeLogger {
  let requestId = '';
  let userId = '';

  const formatMessage = (level: string, message: string, meta?: LogMeta): string => {
    const emoji = getEmoji(level);
    const reqId = requestId ? `[${requestId}]` : '';
    const usr = userId ? `userId: ${userId}` : '';
    const metaStr = meta ? ` | ${JSON.stringify(sanitizeMeta(meta))}` : '';
    const context = [reqId, usr].filter(Boolean).join(' | ');
    
    return `${emoji} [${functionName}] ${message}${context ? ' | ' + context : ''}${metaStr}`;
  };

  const getEmoji = (level: string): string => {
    switch (level) {
      case 'debug': return '🐛';
      case 'info': return 'ℹ️';
      case 'warn': return '⚠️';
      case 'error': return '❌';
      default: return '📝';
    }
  };

  const sanitizeMeta = (meta: LogMeta): LogMeta => {
    const sanitized = { ...meta };
    
    // Remove sensitive fields
    delete sanitized.password;
    delete sanitized.access_token;
    delete sanitized.refresh_token;
    delete sanitized.api_key;
    delete sanitized.secret;
    delete sanitized.card_number;
    delete sanitized.cvv;
    
    // Mask email if present
    if (sanitized.email) {
      sanitized.email = maskEmail(sanitized.email);
    }
    
    // Mask token if present
    if (sanitized.token) {
      sanitized.token = maskToken(sanitized.token);
    }
    
    return sanitized;
  };

  const maskEmail = (email: string): string => {
    try {
      const parts = email.split('@');
      if (parts.length !== 2) return '***';
      
      const localPart = parts[0];
      const domainPart = parts[1];
      
      const maskedLocal = localPart.length > 3 
        ? `${localPart.substring(0, 3)}***`
        : '***';
      
      const maskedDomain = domainPart.length > 3
        ? `${domainPart.substring(0, 3)}***`
        : '***';
      
      return `${maskedLocal}@${maskedDomain}`;
    } catch {
      return '***';
    }
  };

  const maskToken = (token: string): string => {
    if (token.length <= 20) return '***';
    return `${token.substring(0, 10)}...${token.substring(token.length - 10)}`;
  };

  return {
    debug: (message: string, meta?: LogMeta) => {
      console.log(formatMessage('debug', message, meta));
    },
    info: (message: string, meta?: LogMeta) => {
      console.log(formatMessage('info', message, meta));
    },
    warn: (message: string, meta?: LogMeta) => {
      console.warn(formatMessage('warn', message, meta));
    },
    error: (message: string, meta?: LogMeta) => {
      console.error(formatMessage('error', message, meta));
    },
    setRequestId: (id: string) => {
      requestId = id;
    },
    setUserId: (id: string) => {
      userId = id;
    },
  };
}

/**
 * Helper to log RLS policy checks
 */
export function logRLSCheck(
  logger: EdgeLogger,
  tableName: string,
  operation: 'SELECT' | 'INSERT' | 'UPDATE' | 'DELETE',
  userId: string,
): void {
  logger.debug(`🔍 RLS: ${operation} on "${tableName}" for userId: ${userId}`);
}

/**
 * Helper to log database query results
 */
export function logQueryResult(
  logger: EdgeLogger,
  tableName: string,
  operation: 'SELECT' | 'INSERT' | 'UPDATE' | 'DELETE',
  rowCount: number,
  error?: { code: string; message: string },
): void {
  if (error) {
    if (error.code === '42501') {
      logger.error(`🚫 RLS: Permission denied on ${operation} "${tableName}"`, { 
        code: error.code, 
        message: error.message 
      });
    } else {
      logger.error(`❌ DB: ${operation} on "${tableName}" failed`, { 
        code: error.code, 
        message: error.message 
      });
    }
  } else {
    logger.debug(`✅ DB: ${operation} on "${tableName}" successful | rows: ${rowCount}`);
  }
}

File: lib/core/logger.dart
import 'package:logger/logger.dart';
import 'package:flutter/foundation.dart';

/// Centralized logger for the Akeli Nutrition App
/// 
/// This logger provides consistent, structured logging across the application
/// with emoji indicators for different operation types and automatic log level
/// management based on debug/release mode.
/// 
/// ## Usage
/// 
/// ```dart
/// // Import this file
/// import 'package:akeli/core/logger.dart';
/// 
/// // Create a logger instance in your class
/// class MyProvider extends AsyncNotifier<List<Item>> {
///   final _logger = appLogger;
///   
///   @override
///   Future<List<Item>> build() async {
///     _logger.d('🔄 Provider: MyProvider initialized');
///     return _fetchItems();
///   }
/// }
/// ```
/// 
/// ## Log Levels
/// 
/// - `trace`: Extremely detailed (every state change, variable values)
/// - `debug`: Developer debugging (query details, provider lifecycle)
/// - `info`: Important events (auth success, user actions)
/// - `warning`: Potential issues (deprecated API, near limits, RLS blocks)
/// - `error`: Actual errors (exceptions, failed operations)
/// 
/// ## Emoji Legend
/// 
/// | Emoji | Meaning |
/// |-------|---------|
/// | 🔐 | Authentication |
/// | 📡 | Database query |
/// | 🔍 | RLS check |
/// | 🔄 | Provider lifecycle |
/// | ⚡ | Edge function |
/// | 🎯 | User action |
/// | ✅ | Success |
/// | ❌ | Error |
/// | ⚠️ | Warning |
/// | 🚫 | Blocked/Denied |
/// | 💥 | Critical failure |
/// | 👤 | User context |
/// | 🗑️ | Cleanup/Dispose |

class AkeliLogger {
  static final Logger _instance = Logger(
    printer: PrettyPrinter(
      methodCount: 2,
      errorMethodCount: 8,
      lineLength: 120,
      colors: true,
      printEmojis: true,
      printTime: true,
      dateTimeFormat: DateTimeFormat.dateAndTime,
    ),
    level: kDebugMode ? Level.trace : Level.warning,
  );

  static Logger get instance => _instance;

  /// Private constructor to prevent instantiation
  AkeliLogger._();
}

/// Global logger instance for easy access throughout the app
final appLogger = AkeliLogger.instance;

/// Extension methods for category-specific logging
extension AkeliLoggerExtension on Logger {
  // ==================== Authentication Logs ====================

  /// Log authentication events (sign-in, sign-up, sign-out, token refresh)
  void auth(String message, {dynamic error, StackTrace? stackTrace}) {
    if (error != null) {
      e('🔐 Auth: $message', error: error, stackTrace: stackTrace);
    } else {
      i('🔐 Auth: $message');
    }
  }

  // ==================== Database Logs ====================

  /// Log database operations (queries, inserts, updates, deletes)
  void db(String message, {dynamic error, StackTrace? stackTrace}) {
    if (error != null) {
      e('📡 DB: $message', error: error, stackTrace: stackTrace);
    } else {
      d('📡 DB: $message');
    }
  }

  // ==================== RLS Logs ====================

  /// Log Row Level Security checks and violations
  void rls(String message, {dynamic error, StackTrace? stackTrace}) {
    if (error != null) {
      e('🚫 RLS: $message', error: error, stackTrace: stackTrace);
    } else {
      w('🔍 RLS: $message');
    }
  }

  // ==================== Provider Lifecycle Logs ====================

  /// Log Riverpod provider lifecycle events (init, rebuild, dispose)
  void provider(String message, {dynamic error, StackTrace? stackTrace}) {
    if (error != null) {
      e('🔄 Provider: $message', error: error, stackTrace: stackTrace);
    } else {
      d('🔄 Provider: $message');
    }
  }

  // ==================== Edge Function Logs ====================

  /// Log edge function invocations and results
  void edge(String functionName, String message, {dynamic error, StackTrace? stackTrace}) {
    if (error != null) {
      e('⚡ Function: $functionName - $message', error: error, stackTrace: stackTrace);
    } else {
      i('⚡ Function: $functionName - $message');
    }
  }

  // ==================== User Action Logs ====================

  /// Log user actions (button taps, form submissions, gestures)
  void userAction(String action, {String? screen, dynamic metadata}) {
    final screenContext = screen != null ? ' [screen: $screen]' : '';
    final metaStr = metadata != null ? ' | metadata: $metadata' : '';
    i('🎯 UI: $action$screenContext$metaStr');
  }

  // ==================== Navigation Logs ====================

  /// Log navigation/routing events (route changes, auth guards, deep links)
  void navigation(String from, String to, {String? reason}) {
    final reasonStr = reason != null ? ' [reason: $reason]' : '';
    d('🧭 Navigation: $from → $to$reasonStr');
  }

  // ==================== Performance Logs ====================

  /// Log performance metrics (query duration, operation timing)
  void performance(String operation, Duration duration, {String? context}) {
    final contextStr = context != null ? ' | $context' : '';
    if (duration.inMilliseconds > 1000) {
      w('⏱️ Perf: $operation took ${duration.inMilliseconds}ms$contextStr');
    } else {
      d('⏱️ Perf: $operation took ${duration.inMilliseconds}ms$contextStr');
    }
  }
}

/// Helper function to mask sensitive data in logs
class LogHelper {
  /// Mask email address (show first 3 chars and domain first 3 chars)
  /// Example: "john.doe@example.com" → "joh***@exa***"
  static String maskEmail(String email) {
    try {
      final parts = email.split('@');
      if (parts.length != 2) return '***';
      
      final localPart = parts[0];
      final domainPart = parts[1];
      
      final maskedLocal = localPart.length > 3 
          ? '${localPart.substring(0, 3)}***'
          : '***';
      
      final maskedDomain = domainPart.length > 3
          ? '${domainPart.substring(0, 3)}***'
          : '***';
      
      return '$maskedLocal@$maskedDomain';
    } catch (e) {
      return '***';
    }
  }

  /// Mask UUID (show first 4 and last 4 chars)
  /// Example: "abc123-def456-ghi789" → "abc1***h789"
  static String maskUuid(String uuid) {
    if (uuid.length <= 8) return '***';
    return '${uuid.substring(0, 4)}***${uuid.substring(uuid.length - 4)}';
  }

  /// Mask token (show first 10 and last 10 chars)
  static String maskToken(String token) {
    if (token.length <= 20) return '***';
    return '${token.substring(0, 10)}...${token.substring(token.length - 10)}';
  }

  /// Sanitize object for safe logging (removes sensitive fields)
  static Map<String, dynamic> sanitizeData(Map<String, dynamic> data) {
    final sanitized = Map<String, dynamic>.from(data);
    
    // Remove sensitive fields
    sanitized.remove('password');
    sanitized.remove('access_token');
    sanitized.remove('refresh_token');
    sanitized.remove('api_key');
    sanitized.remove('secret');
    sanitized.remove('card_number');
    sanitized.remove('cvv');
    
    // Mask email if present
    if (sanitized.containsKey('email')) {
      sanitized['email'] = maskEmail(sanitized['email']);
    }
    
    // Mask token if present
    if (sanitized.containsKey('token')) {
      sanitized['token'] = maskToken(sanitized['token']);
    }
    
    return sanitized;
  }
}

/// RLS Debug Helper
/// Use this when you suspect RLS is blocking access
class RLSDebugHelper {
  /// Log detailed RLS context when queries return unexpected results
  static void debugQuery(
    String tableName,
    String? userId, {
    Map<String, dynamic>? filters,
    int rowCount = 0,
  }) {
    appLogger.d('🔍 RLS DEBUG: Querying "$tableName" table');
    appLogger.d('  userId: ${userId ?? "null (not authenticated)"}');
    appLogger.d('  filters: ${filters ?? "none"}');
    appLogger.d('  rows returned: $rowCount');
    
    if (rowCount == 0 && userId != null) {
      appLogger.w('⚠️ RLS DEBUG: Possible RLS policy blocking userId: $userId');
      appLogger.w('  Check policies on "$tableName" table for auth_uid() match');
    }
  }

  /// Log RLS policy check result
  static void logPolicyCheck(
    String tableName,
    String policyName,
    bool allowed, {
    String? userId,
  }) {
    if (allowed) {
      appLogger.d('✅ RLS: Policy "$policyName" on "$tableName" allowed for userId: $userId');
    } else {
      appLogger.w('🚫 RLS: Policy "$policyName" on "$tableName" blocked for userId: $userId');
    }
  }
}
